compiler/ir/backend.wasm/src/org/jetbrains/kotlin/backend/wasm/ir2wasm/DeclarationGenerator.kt

```diff
@@ -0,0 +1,329 @@
+                // will be stored Wasm table and could be called indirectly.
+                context.registerVirtualFunction(declaration.symbol)
+            }
+
+            if (!isVirtual && isIntrinsic) {
+                // Calls to non-virtual intrinsic functions are replaced with something else.
+                // No need to generate them.
+                return
+            }
+        }
+
+        assert(declaration == declaration.realOverrideTarget) {
+            "Sanity check that $declaration is a real function that can be used in calls"
+        }
+
+        if (importedName != null) {
+            // Imported functions don't have bodies. Declaring the signature:
+            context.defineFunction(
+                declaration.symbol,
+                WasmFunction.Imported(watName, wasmFunctionType, importedName)
+            )
+            // TODO: Support re-export of imported functions.
+            return
+        }
+
+        val function = WasmFunction.Defined(watName, wasmFunctionType)
+        val functionCodegenContext = WasmFunctionCodegenContextImpl(
+            declaration,
+            function,
+            backendContext,
+            context
+        )
+
+        for (irParameter in irParameters) {
+            functionCodegenContext.defineLocal(irParameter.symbol)
+        }
+
+        val exprGen = functionCodegenContext.bodyGen
+        val bodyBuilder = BodyGenerator(functionCodegenContext)
+
+        if (isIntrinsic) {
+            bodyBuilder.tryToGenerateWasmOpIntrinsicCall(declaration)
+        } else {
+            when (val body = declaration.body) {
+                is IrBlockBody ->
+                    for (statement in body.statements) {
+                        bodyBuilder.statementToWasmInstruction(statement)
+                    }
+
+                is IrExpressionBody ->
+                    bodyBuilder.generateExpression(body.expression)
+
+                else -> error("Unexpected body $body")
+            }
+        }
+
+        // Return implicit this from constructions to avoid extra tmp
+        // variables on constructor call sites.
+        // TODO: Redesign construction scheme.
+        if (declaration is IrConstructor) {
+            exprGen.buildGetLocal(/*implicit this*/ function.locals[0])
+            exprGen.buildInstr(WasmOp.RETURN)
+        }
+
+        // Add unreachable if function returns something but not as a last instruction.
+        if (wasmFunctionType.resultTypes.isNotEmpty() && declaration.body is IrBlockBody) {
+            exprGen.buildUnreachable()
+        }
+
+        context.defineFunction(declaration.symbol, function)
+
+        if (declaration == backendContext.startFunction)
+            context.setStartFunction(function)
+
+        if (declaration.isExported(backendContext)) {
+            context.addExport(
+                WasmExport.Function(
+                    field = function,
+                    name = declaration.name.identifier
```

Comment:
TODO: add an ability to define export name

Reply:
TODO added